### PR TITLE
feat: support for `Server Deactivate All PDU`

### PR DESCRIPTION
**Pull request by ibeckermayer**
The [Deactivate All PDU](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpbcgr/8a29971a-df3c-48da-add2-8ed9a05edc89)  is sent from server to client to indicate that the connection will be dropped or that a capability re-exchange using a [Deactivation-Reactivation Sequence](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpbcgr/dfc234ce-481a-4674-9a5d-2a7bafb14432) will occur.

In practice, this happens when a screen resize is requested. When this PDU is received by the client, it is meant to begin executing the Deactivation-Reactivation Sequence, which is the Capability Exchange and Connection Finalization phases of the connection sequence.

What this PR does is

1. Break out the Capability Exchange and Connection Finalization phases of the connection sequence into their own logical abstraction `ConnectionActivationSequence`. This allows this execution `Sequence` to be used both during the initial connection phase, as well as be passed out as an [`x224::ProcessorOutput`](https://github.com/Devolutions/IronRDP/blob/717d02068b2cbcae804c3a5043d4f440a8ae62d8/crates/ironrdp-session/src/x224/mod.rs#L33-L37) in the event of a Deactivate All PDU.
2. Refactors the [`rfx::DecodingContext`](https://github.com/Devolutions/IronRDP/blob/717d02068b2cbcae804c3a5043d4f440a8ae62d8/crates/ironrdp-session/src/rfx.rs#L17-L21) to remove its explicit state management. This is included because, upon a successful Deactivation-Reactivation Sequence executing, the remote fx system is also reset, with a Sync being re-sent and so on. With this refactor, rather than needing to communicate this "reset" from the x224 processor over to the fast path processor and ultimately the `rfx::DecodingContext` to update its explicit [`state`](https://github.com/Devolutions/IronRDP/blob/master/crates/ironrdp-session/src/rfx.rs#L18) to the right value, we just let the server keep track of all state and simply decode based on whatever header it sends us.

In terms of existing IronRDP clients (like web and native), these changes should have no effect, since they are not hooked up to send screen resizes yet. This was tested via wiring things up in Teleport which can be seen in part starting [here](https://github.com/gravitational/teleport/blob/92929e381f95526cf7e0a2ef39462413f5fedfef/lib/srv/desktop/rdp/rdpclient/src/client.rs#L329).

### CODE REVIEW

**Comment by pacmancoder** [comment]
This looks good to me 🚀 ! Let's wait for @CBenoit and merge this to unblock further PRs

PS: Clippy lints are failing (`xtask check lints`)

**Comment by elmarco** [comment]
lgtm overall, however, please make it compile and run test after each commit!

File: crates/ironrdp-web/src/session.rs
Comment (CBenoit) [inline crates/ironrdp-web/src/session.rs:606]: issue: Can’t we handle this in a more graceful manner?

File: Cargo.toml
Comment (CBenoit) [inline Cargo.toml:62]: note: This is a legacy dependency, now that we have ReadCursor

Comment (ibeckermayer) [reply Cargo.toml:62]: These are marked as deprecated in the follow up PR https://github.com/Devolutions/IronRDP/pull/419/commits/9503901f066b2341480c1a2b4a35357d8c262944.

File: crates/ironrdp-async/src/connector.rs
Comment (CBenoit) [inline crates/ironrdp-async/src/connector.rs:194]: question: Is it intended to be public?

File: crates/ironrdp-pdu/src/codecs/rfx/data_messages.rs
Comment (CBenoit) [inline crates/ironrdp-pdu/src/codecs/rfx/data_messages.rs:138]: question: What is this trait? It seems to be a legacy one.

**Comment by CBenoit** [comment]
Can you rebase on top of master to get the fuzzing fixes?

File: crates/ironrdp-async/src/connector.rs
Comment (ibeckermayer) [reply crates/ironrdp-async/src/connector.rs:194]: It's used by Teleport for executing the deactivation-reactivation sequence. It appears to be a generally useful pattern, since I also needed it the [attempted] implementation for IronRDP web: https://github.com/Devolutions/IronRDP/pull/421/files#diff-33dc7055fcffcb2318664572070611527a75668ad24c01e2ce01ae83e0b27defR616

File: crates/ironrdp-pdu/src/codecs/rfx/data_messages.rs
Comment (ibeckermayer) [reply crates/ironrdp-pdu/src/codecs/rfx/data_messages.rs:138]: I don't know anything more than you about the trait, however it looks like it's defined in a `legacy` module and is still heavily used:

```
$ grep -rn 'impl PduBufferParsing'.
crates/ironrdp-pdu/src/codecs/rfx/data_messages.rs:78:impl PduBufferParsing<'_> for ContextPdu {
crates/ironrdp-pdu/src/codecs/rfx/data_messages.rs:138:impl PduBufferParsing<'_> for FrameBeginPdu {
crates/ironrdp-pdu/src/codecs/rfx/data_messages.rs:170:impl PduBufferParsing<'_> for FrameEndPdu {
crates/ironrdp-pdu/src/codecs/rfx/data_messages.rs:203:impl PduBufferParsing<'_> for RegionPdu {
crates/ironrdp-pdu/src/codecs/rfx/data_messages.rs:424:impl PduBufferParsing<'_> for RfxRectangle {
crates/ironrdp-pdu/src/codecs/rfx/data_messages.rs:464:impl PduBufferParsing<'_> for Quant {
crates/ironrdp-pdu/src/codecs/rfx/header_messages.rs:36:impl PduBufferParsing<'_> for SyncPdu {
crates/ironrdp-pdu/src/codecs/rfx/header_messages.rs:80:impl PduBufferParsing<'_> for CodecVersionsPdu {
crates/ironrdp-pdu/src/codecs/rfx/header_messages.rs:130:impl PduBufferParsing<'_> for ChannelsPdu {
crates/ironrdp-pdu/src/codecs/rfx/header_messages.rs:211:impl PduBufferParsing<'_> for RfxChannel {
crates/ironrdp-pdu/src/codecs/rfx/header_messages.rs:245:impl PduBufferParsing<'_> for CodecVersion {
```

File: crates/ironrdp-web/src/session.rs
Comment (ibeckermayer) [reply crates/ironrdp-web/src/session.rs:606]: Presuming [this system](https://github.com/Devolutions/IronRDP/pull/421/files#diff-33dc7055fcffcb2318664572070611527a75668ad24c01e2ce01ae83e0b27defR609-R655) is confirmed to work in web (just waiting for another test from Nefix in matrix), I think I can safely copy that into here.

File: crates/ironrdp-pdu/src/rdp/headers.rs
Comment (CBenoit) [inline crates/ironrdp-pdu/src/rdp/headers.rs:527]: issue: `ensure_fixed_part_size!(in: src);` / `ensure_size!` missing (the fuzzer is failing because of this)

File: crates/ironrdp-pdu/src/rdp/headers.rs
Comment (CBenoit) [inline crates/ironrdp-pdu/src/rdp/headers.rs:533]: issue:
```suggestion
        ensure_size!(in: dst, size: self.size());

        // A 16-bit, unsigned integer. The size in bytes of the sourceDescriptor field.
```

**Comment by pacmancoder** [comment]
@ibeckermayer Looks like there is a problem found by fuzzing in `ShareControlPdu` pdu changes -- the following input for `pdu_decode` does not fail in master, but fails in branch: `[0, 4, 22, 0, 5, 0, 0, 0, 5, 0, 5, 40]` (see fuzzing logs)

You could just temporarily add a test in `crates\ironrdp-fuzzing\src\oracles\mod.rs` and investigate it from here

```rust
#[test]
fn test_pdu_decode() {
    const DATA: &'static [u8] = &[0, 4, 22, 0, 5, 0, 0, 0, 5, 0, 5, 40];
    pdu_decode(DATA);
}
```

cc @CBenoit

File: crates/ironrdp-pdu/src/rdp/headers.rs
Comment (pacmancoder) [reply crates/ironrdp-pdu/src/rdp/headers.rs:533]: Most likely, this causing fuzzing fo fail

File: crates/ironrdp-async/src/connector.rs
Comment (CBenoit) [inline crates/ironrdp-async/src/connector.rs:194]: naming: If I understand correctly, it’s not only about "connecting" anymore, am I right? It’s now a general-purpose helper for anything that is a `Sequence`. I suggest to move that into the `framed` module and rename to a more appropriate name. You can do the same for `single_connect_step_write`.

**Review by CBenoit — COMMENTED** [review]
Looking good to me! Can be merged once remaining comments are addressed :+1:

File: crates/ironrdp-pdu/src/rdp/headers.rs
Comment (CBenoit) [inline crates/ironrdp-pdu/src/rdp/headers.rs:521]: style (non-blocking): The style used across the codebase is the shorthand style (without `{}`)

File: crates/ironrdp-async/src/connector.rs
Comment (ibeckermayer) [reply crates/ironrdp-async/src/connector.rs]: See db346b3c2aaf2be522b55d67a595f88c62cc5b29

File: crates/ironrdp-pdu/src/rdp/headers.rs
Comment (ibeckermayer) [reply crates/ironrdp-pdu/src/rdp/headers.rs:532]: 3cd8167729697ab85a9674d347829e010bd4c566

File: crates/ironrdp-pdu/src/rdp/headers.rs
Comment (ibeckermayer) [reply crates/ironrdp-pdu/src/rdp/headers.rs]: Oops, didn't mean to change that: 1de16664c6c7e90b0f9c5d9c3efc716f9bdbd631

**Review by pacmancoder — APPROVED** [review]
LGTM!

**Pull request merged**
